### PR TITLE
Add RAII-based API for recording events and bump version to 0.4.0

### PR DESCRIPTION
**Pull request by michaelwoerister**
This PR adds a RAII-based API for recording events in preparation for replacing `(start, end)` event pairs with single "interval" events. It also bumps the version to 0.4.0 so the new API can be used in `rustc`. 

r? @wesleywiser

### CODE REVIEW

File: measureme/src/profiler.rs [hunk 1]
```diff
@@ -111,4 +113,46 @@ impl<S: SerializationSink> Profiler<S> {
                 bytes.copy_from_slice(raw_event_bytes);
             });
     }
+
+    /// Creates a "start" event and returns a `TimingGuard` that will create
+    /// the corresponding "end" event when it is dropped.
+    pub fn start_event<'a>(
```
Comment (michaelwoerister) [inline measureme/src/profiler.rs]: Is there a better name for this? `start_recording` maybe?

File: measureme/src/profiler.rs [hunk 2]
```diff
@@ -111,4 +113,46 @@ impl<S: SerializationSink> Profiler<S> {
...
+            event_id,
+            event_kind,
+            thread_id,
+        })
+    }
+}
+
+/// When dropped, this `TimingGuard` will record an "end" event in the
+/// `Profiler` it was created by.
+#[must_use]
```
Comment (bjorn3) [inline measureme/src/profiler.rs]: Why must_use? That would mean that you would have to manually call `std::mem::drop` to prevent a compiler warning.

Comment (michaelwoerister) [reply measureme/src/profiler.rs]: `let _guard = ...` also works without a warning. This code is basically copied from what we already have in `rustc`:
https://github.com/rust-lang/rust/blob/8e0007f829661e57d008d2e908c95f6e84b04b25/src/librustc/util/profiling.rs#L348-L349

It works well there.

Comment (michaelwoerister) [reply measureme/src/profiler.rs]: And in the compiler version the `#[must_use]` attribute has prevented me from introducing at least two bugs, so I think it's useful `:)`

Comment (bjorn3) [reply measureme/src/profiler.rs]: Didnt know that.

File: measureme/src/profiler.rs (hunk 1 above)
Comment (michaelwoerister) [reply measureme/src/profiler.rs]: Is `start_recording_interval_event` too wordy?

Comment (wesleywiser) [reply measureme/src/profiler.rs]: `start_recording` implies to me that if you don't call it, any calls to `record_event` will not be recorded.

I think this is fine for now.

File: measureme/src/profiler.rs [hunk 3]
```diff
@@ -111,4 +113,46 @@ impl<S: SerializationSink> Profiler<S> {
...
+pub struct TimingGuard<'a, S: SerializationSink>(TimingGuardInternal<'a, S>);
+
+struct TimingGuardInternal<'a, S: SerializationSink> {
```
Comment (michaelwoerister) [inline measureme/src/profiler.rs]: I just realized that the `TimingGuard`/`TimingGuardInternal` split is not necessary for measureme. Let me fix that...

File: measureme/src/profiler.rs (hunk 1 above)
Comment (wesleywiser) [reply measureme/src/profiler.rs]: We could eventually split the api more and hide `record_event` and just expose two different methods `record_duration` and `record_instantaneous`.

**Comment by wesleywiser** [comment]
Are there changes we should make to `RawEvent` before we ship 0.4?

File: measureme/src/profiler.rs (hunk 1 above)
Comment (michaelwoerister) [reply measureme/src/profiler.rs]: I changed it to `start_recording_interval_event` just now, and plan to later replace `record_event` with `record_instant_event`. Does that work for you?

**Comment by michaelwoerister** [comment]
> Are there changes we should make to `RawEvent` before we ship 0.4?

We could change ThreadId to `u32` but the rest of the changes depend on actually switching to "interval" events, I'd say.

**Comment by michaelwoerister** [comment]
But let's modify `RawEvent` just now. None of the changes above affect `summarize` or other tools, so we can bump `rustc` to 0.4.0 without worrying about perf.rlo just now.

File: measureme/src/profiler.rs (hunk 1 above)
Comment (wesleywiser) [reply measureme/src/profiler.rs]: Yeah, that's fine.

**Comment by wesleywiser** [comment]
> But let's not modify RawEvent just yet. None of the changes above affect summarize or other tools, so we can bump rustc to 0.4.0 without worrying about perf.rlo.

I agree. When we get to that point, we'll have to make sure we loop in @Mark-Simulacrum so we don't break perf.rlo.
